packages/databyss-ui/components/PageContent/PageBody.js

```diff
@@ -1,29 +1,55 @@
-import React, { useEffect } from 'react'
-import EditorProvider, {
-  pageReducer,
-} from '@databyss-org/ui/editor/EditorProvider'
+import React, { useEffect, useCallback, useRef } from 'react'
+import { throttle } from 'lodash'
+import ContentEditable from '@databyss-org/editor/components/ContentEditable'
+import EditorProvider from '@databyss-org/editor/state/EditorProvider'
+import { withMetaData } from '@databyss-org/editor/lib/util'
 import { usePageContext } from '@databyss-org/services/pages/PageProvider'
 import { useNavigationContext } from '@databyss-org/ui'
-import AutoSave from '@databyss-org/ui/editor/AutoSave'
-import EditorPage from '@databyss-org/ui/editor/EditorPage'
-import SlateContentEditable from '@databyss-org/ui/editor/slate/page/ContentEditable'
-import slateReducer from '@databyss-org/ui/editor/slate/page/reducer'
+import {
+  withWhitelist,
+  addMetaData,
+} from '@databyss-org/services/pages/_helpers'
 
-const PageBody = ({ page, readOnly }) => {
+const PageBody = ({ page }) => {
   const { location } = useNavigationContext()
-  const { clearBlockDict } = usePageContext()
+  const { clearBlockDict, setPatch } = usePageContext()
   useEffect(() => () => clearBlockDict(), [])
+
+  const operationsQueue = useRef([])
+
+  // throttled autosave occurs every SAVE_PAGE_THROTTLE ms when changes are happening
+  const throttledAutosave = useCallback(
+    throttle(({ state, patch }) => {
+      const _patch = withWhitelist(patch)
+      if (_patch.length) {
+        const payload = {
+          id: state.page._id,
+          patch: operationsQueue.current,
+        }
+        setPatch(payload)
+        operationsQueue.current = []
+      }
+    }, process.env.SAVE_PAGE_THROTTLE),
+    []
+  )
+
+  // TODO: Not all events fire onchange even in editor
```

Comment:
Like which events?

Reply:
this has been fixed by moving the `onChange` function on `ContentEditable`